[BUG]直播间重新开播，直播流未正常进行录制

**描述：**
25-3-22 录播遇到一次主播直播间关闭，后开播后发现未正常录制后续内容

**如何遇到：**
加入直播间，启用自动录播，默认设置（除了改了缓存和输出文件夹）

**期望：**
日志没找到那一时间段的提示，不清楚什么问题，期望作者解答下

**日志和截图：**
见下方代码段（2点多电脑休眠了）
```
16:57:46 [INFO] Open player window: 23771139 1742644931
03:38:28 [ERROR] Failed fetching index content from https://d1--cn-gotcha204.bilivideo.com/live-bvc/668524/live_1878154667_69909417_bluray/index.m3u8?expires=1742663053&len=0&oi=2871785903&pt=h5&qn=10000&trid=1007202280194af7eca66c8d00710767dedf&sigparams=cdn,expires,len,oi,pt,qn,trid&cdn=cn-gotcha204&sign=47ae925fa1d855d54ac68441757b74ad&site=79efd27f93a6765e81bed2d3a3da91a4&free_type=0&mid=3546873990219945&sche=ban&bvchls=1&trace=16&isp=ct&rg=Central&pv=Hubei&pp=rtmp&score=1&sk=5bf07b9bbe6df2e0a6bc476fe3d9a642d4caf7efae2e198b21bcc415a9c6f5c2&p2p_type=-1&origin_bitrate=785339&deploy_env=prod&flvsk=4207df3de646838b084f14f252be3aff3033194328f6029e3e02d13ef8059a83&sl=3&hot_cdn=909769&suffix=bluray&info_source=cache&source=puv3_onetier&vd=bc&src=puv3&order=1
03:38:28 [ERROR] [23771139]Update entries error: BiliClient error: Client error: error sending request for url (https://d1--cn-gotcha204.bilivideo.com/live-bvc/668524/live_1878154667_69909417_bluray/index.m3u8?expires=1742663053&len=0&oi=2871785903&pt=h5&qn=10000&trid=1007202280194af7eca66c8d00710767dedf&sigparams=cdn,expires,len,oi,pt,qn,trid&cdn=cn-gotcha204&sign=47ae925fa1d855d54ac68441757b74ad&site=79efd27f93a6765e81bed2d3a3da91a4&free_type=0&mid=3546873990219945&sche=ban&bvchls=1&trace=16&isp=ct&rg=Central&pv=Hubei&pp=rtmp&score=1&sk=5bf07b9bbe6df2e0a6bc476fe3d9a642d4caf7efae2e198b21bcc415a9c6f5c2&p2p_type=-1&origin_bitrate=785339&deploy_env=prod&flvsk=4207df3de646838b084f14f252be3aff3033194328f6029e3e02d13ef8059a83&sl=3&hot_cdn=909769&suffix=bluray&info_source=cache&source=puv3_onetier&vd=bc&src=puv3&order=1): operation timed out
03:38:28 [WARN] danmu will reconnect after 3s
```

问题大概在 3-23 00：00时间点左右

**相关信息：**
- 程序版本：2.0.2
- 系统类型：x64 win11 Amd

**其他**
其他都挺好的，支持大佬

<img width="833" alt="Image" src="https://github.com/user-attachments/assets/312caa19-0131-4cb6-bfc2-9d2ea3257bd4" />

--- a/src-tauri/src/recorder/bilibili.rs
+++ b/src-tauri/src/recorder/bilibili.rs
@@ -199,24 +199,6 @@ impl BiliRecorder {
                 // if stream is confirmed to be closed, live stream cache is cleaned.
                 // all request will go through fs
                 if live_status {
-                    let mut rng = rand::thread_rng();
-                    // WHY: when program started, all stream is fetched nearly at the same time, so they will expire toggether,
-                    // this might meet server rate limit. So we add a random offset to make request spread over time.
-                    let offset = rng.gen_range(5..=120);
-                    // no need to update stream as it's not expired yet
-                    if self
-                        .live_stream
-                        .read()
-                        .await
-                        .as_ref()
-                        .is_some_and(|s| s.expire - offset > Utc::now().timestamp())
-                    {
-                        return live_status;
-                    }
-                    log::info!(
-                        "[{}]Stream is empty or nearly expired, updating",
-                        self.room_id
-                    );
                     match self
                         .client
                         .read()
@@ -603,6 +585,24 @@ impl BiliRecorder {
                 return Err(e);
             }
         }
+
+        // check stream is nearly expired
+        // WHY: when program started, all stream is fetched nearly at the same time, so they will expire toggether,
+        // this might meet server rate limit. So we add a random offset to make request spread over time.
+        let mut rng = rand::thread_rng();
+        let pre_offset = rng.gen_range(5..=120);
+        // no need to update stream as it's not expired yet
+        let current_stream = self.live_stream.read().await.clone();
+        if current_stream
+            .as_ref()
+            .is_some_and(|s| s.expire - Utc::now().timestamp() < pre_offset)
+        {
+            log::info!("Stream is nearly expired");
+            return Err(super::errors::RecorderError::StreamExpired {
+                stream: current_stream.unwrap(),
+            });
+        }
+
         Ok(task_begin_time.elapsed().as_millis())
     }
 

--- a/src-tauri/src/recorder/bilibili/client.rs
+++ b/src-tauri/src/recorder/bilibili/client.rs
@@ -2,16 +2,17 @@ use super::errors::BiliClientError;
 use super::profile;
 use super::profile::Profile;
 use super::response;
-use base64::Engine;
-use pct_str::PctString;
-use pct_str::URIReserved;
-use regex::Regex;
-use reqwest::Client;
 use super::response::Format;
 use super::response::GeneralResponse;
 use super::response::PostVideoMetaResponse;
 use super::response::PreuploadResponse;
 use super::response::VideoSubmitData;
+use crate::database::account::AccountRow;
+use base64::Engine;
+use pct_str::PctString;
+use pct_str::URIReserved;
+use regex::Regex;
+use reqwest::Client;
 use serde::Deserialize;
 use serde::Serialize;
 use serde_json::json;
@@ -23,7 +24,6 @@ use std::time::SystemTime;
 use tokio::fs::File;
 use tokio::io::AsyncReadExt;
 use tokio::time::Instant;
-use crate::database::account::AccountRow;
 
 #[derive(Serialize, Deserialize, Clone, Debug)]
 pub struct RoomInfo {
@@ -322,7 +322,9 @@ impl BiliClient {
         let response = self.client.get(url).send().await?;
         let bytes = response.bytes().await?;
         let base64 = base64::engine::general_purpose::STANDARD.encode(bytes);
-        let mime_type = mime_guess::from_path(url).first_or_octet_stream().to_string();
+        let mime_type = mime_guess::from_path(url)
+            .first_or_octet_stream()
+            .to_string();
         Ok(format!("data:{};base64,{}", mime_type, base64))
     }
 

--- a/src-tauri/src/recorder/errors.rs
+++ b/src-tauri/src/recorder/errors.rs
@@ -1,6 +1,6 @@
-use custom_error::custom_error;
 use super::bilibili::client::BiliStream;
 use super::douyin::client::DouyinClientError;
+use custom_error::custom_error;
 
 custom_error! {pub RecorderError
     IndexNotFound {url: String} = "Index not found: {url}",
@@ -9,6 +9,7 @@ custom_error! {pub RecorderError
     M3u8ParseFailed {content: String } = "Parse m3u8 content failed: {content}",
     NoStreamAvailable = "No available stream provided",
     FreezedStream {stream: BiliStream} = "Stream is freezed: {stream}",
+    StreamExpired {stream: BiliStream} = "Stream is nearly expired: {stream}",
     NoRoomInfo = "No room info provided",
     InvalidStream {stream: BiliStream} = "Invalid stream: {stream}",
     SlowStream {stream: BiliStream} = "Stream is too slow: {stream}",

